### DISCUSSION TITLE
Versioned DLLs can cause broken upgrades in Windows Installer - what do PyInstaller users do?

### DISCUSSION THREAD

**Discussion by SnoopJ** [pyinstaller/pyinstaller#7319] [Help]
I originally posted about this issue [on discuss.python.org](https://discuss.python.org/t/python-or-third-party-dll-file-versions-can-interact-poorly-with-windows-installers-upgrade-functionality-what-to-do/21819) and was suggested to ask this question directly to the PyInstaller community. I wrote a long post over there but I'll try to be more succinct here.

Files on Windows allow defining a `File Version` attribute, and an [unfortunate quirk](https://learn.microsoft.com/en-us/archive/blogs/astebner/why-windows-installer-removes-files-during-a-major-upgrade-if-they-go-backwards-in-version-numbers) of Windows Installer means that if this version goes backwards, the file will be totally missing when using Installer to upgrade the application.

In my case, I have a PyInstaller application that is packaged up with Windows Installer, and I've discovered that some of the files bundled in it set this file attribute. I'm aware of a few options for getting around the issue:

1) Somehow ensure that file versions never go backwards, i.e. avoid having the problem in the first place.

2) Modify the sequencing of the installer so that the pre-upgrade version is completely removed before the post-upgrade files are put in place (i.e. `RemoveExistingProducts` comes before `CostFinalize`). This is technically forbidden by Windows Installer but some users have reported good results from using it anyway, ignoring validation errors/etc.

3) Modify the installer so that file versions are ignored (i.e. setting the `REINSTALLMODE` property in the installer to include the `a` flag), and all files in the newer distribution are selected for installation. I believe this is also forbidden, the relevant property is only supposed to be set at runtime, but it does work. This is the option I used to get around the problem when I discovered it.

4) Modify the `File Version` of the offending files, truncating or removing it entirely. I've found some reference to Windows Installer users doing this, but it seems like it could be risky if any of the software depends on the attribute.

I'm wondering what other users with workflows at the intersection of PyInstaller and Windows Installer do about this problem. Have I missed some options here?

**Comment by bwoodsend** [comment]
Ughh gross, I had no idea that this _feature_ existed.

> The MSI in question has the RemoveExistingProducts action scheduled after the InstallInitialize action (based on this [Windows Installer documentation](https://msdn.microsoft.com/en-us/library/aa371197(v=vs.85).aspx)), and Windows Installer evaluates whether or not it needs to install each of the components in version 2 of the MSI before it uninstalls the components in version 1.  Because a higher version of the component’s key path file is found on the PC, it decides that it needs to skip attempting to install the component.  However, after that decision has been made to skip attempting to install the component, the component is then removed during the RemoveExistingProducts action.

Honestly, I'd call that a design bug in Windows Installer to me. It should exclude the libraries it's about to remove from its search. There should be no difference between an upgrade and a clean uninstall/reinstall.

If Windows is going to foolishly attempt sharing dependencies and library versioning (i.e. highly nontrivial package management) without a package manager whenever it encounters libraries with `File Version` set then stripping `File Version` from each library sounds like the least destructive option to me. It'll invalidate the signature on `python3.dll` so you'll have to either strip that too or resign everything but, whilst that's an extra step, you're not really losing anything since the rest of your application won't be signed.
